### PR TITLE
Configure cors

### LINKED ISSUE #65: Configure cors
* Only allowing `cors` from the CompE+ frontend
* Allowing pre-flight OPTION requests

### PR DESCRIPTION
**Pull request by armiantos**
# Overview

* Specify allowed origins header

# Changes

* Configure `cors` as mentioned in [Configuring CORS w/ Dynamic Origin](https://www.npmjs.com/package/cors#configuring-cors-w-dynamic-origin)

# Questions & Notes

* Do we want to fail a request if it doesn't have the origin tag? Cors wasn't designed for security so I personally think it should not throw an error
* This doesn't support preview builds yet because they have funky origin names and we might need regex to handle that

# Issue tracking

Closes #65

# Testing & Demo

Refer to unit tests

### CODE REVIEW

**Review by calebschoepp — CHANGES_REQUESTED** [review]
I agree that origin should not be required.

Maybe file a ticket to support preview urls with cors.

Thanks for doing this.

File: api/src/controllers/middleware.test.ts [hunk 1]
```diff
@@ -183,3 +184,45 @@ describe('errorHandler middleware', () => {
     // Can't figure out how to mock req.log.error so I'm punting on this
     it.todo('handles unknown errors');
 });
+
+describe('cors middleware', () => {
+    const allowedOrigin = 'http://example.com';
+    const allowedOrigins = [allowedOrigin];
+
+    const corsMiddleware = cors({
```
[the remark this answers is not in the document]
Comment (armiantos) [reply api/src/controllers/middleware.test.ts:192]: Agreed, I'll move it over

File: api/src/server.ts [hunk 2]
```diff
@@ -11,12 +11,23 @@ import logger from './util/logger';
...
+/** Allow cross origin requests */
+router.use(
+    cors({
+        origin: (origin, callback) => {
+            if (allowedOrigins.some((allowedOrigin) => allowedOrigin === origin)) {
+                callback(null, true);
+            }
+            callback(null);
+        },
+    }),
```
[the remark this answers is not in the document]
Comment (armiantos) [reply api/src/server.ts:28]: Yes!

**Comment by armiantos** [comment]
Filed #176 to support url wildcards

File: api/src/controllers/middleware.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (armiantos) [reply api/src/controllers/middleware.test.ts]: Moved, let me know if there's anything else I should update

**Review by calebschoepp — APPROVED** [review]
LGTM! :ship: :smile:

**Pull request merged**